src/core/router/identity.cc

```diff
@@ -730,27 +730,23 @@ Keys CreateRandomKeys() {
 IdentHash CreateRoutingKey(
     const IdentHash& ident) {
   std::uint8_t buf[41];  // ident + yyyymmdd
-  memcpy(buf, (const std::uint8_t *)ident, 32);
+  std::memcpy(buf, (const std::uint8_t *)ident, 32);
   time_t t = time(nullptr);
   struct tm tm;
-  // TODO(unassigned): never use sprintf, use snprintf instead.
+  std::stringstream ss;
+  ss.imbue(std::locale(std::locale::classic()));
 #ifdef _WIN32
-  gmtime_s(&tm, &t);
-  sprintf_s(
-      reinterpret_cast<char *>((buf + 32)),
-      9,
-      "%04i%02i%02i",
-      tm.tm_year + 1900,
-      tm.tm_mon + 1,
-      tm.tm_mday);
+  gmtime_s(&t, &tm);
+  tm.tm_year += 1900;
+  tm.tm_mon += 1;
+  ss << std::put_time(&tm, "%Y%m%d");
+  std::memcpy(reinterpret_cast<char *>((buf + 32)), ss.str().data(), 9);
 #else
   gmtime_r(&t, &tm);
-  sprintf(
-      reinterpret_cast<char *>((buf + 32)),
-      "%04i%02i%02i",
-      tm.tm_year + 1900,
-      tm.tm_mon + 1,
-      tm.tm_mday);
+  tm.tm_year += 1900;
+  tm.tm_mon += 1;
+  ss << std::put_time(&tm, "%Y%m%d");
+  std::memcpy(reinterpret_cast<char *>((buf + 32)), ss.str().data(), 9);
```

Comment:
I don't get 1. why this is not done with [std::strftime](http://en.cppreference.com/w/cpp/chrono/c/strftime)
2. why platform specific ``gmtime_r/s`` ?

What is wrong with:
```
  std::time_t current_time = std::time(nullptr);
  char time_string[9] = {};
  auto ret = std::strftime(time_string, sizeof(time_string),
                           "%Y%m%d", std::gmtime(&current_time));
```

This should work for both platforms.

Reply:
I see that ``std::gmtime`` may not be thread-safe.
``std::gmtime_s`` is C11 standard alternative for POSIX ``gmtime_r``.

Reply:
> why this is not done with std::strftime

Good question :). I wasn't familiar with the function, and must have missed it when looking at the reference material.

> I see that std::gmtime may not be thread-safe.

Originally, I had left those alternatives in for this reason, and will add them back. Thanks for the review @rakhimov.

Reply:
After looking at the documentation, `std::put_time` makes more sense for use with streams. If arrays and `memcpy` are better, then I'll use your suggestion.

Ideally, I would like to get rid of the `gmtime_*` calls as well. Boost.DateTime seems the only viable thread-safe alternative, but looks like a lot of overhead for such a simple use-case. Maybe it's worth it...

Reply:
Please consider doing whatever it takes to present a clean, streamlined, simple, efficient implementation of this function. A simple cross-platform thread-safe solution is welcome.

Reply:
Will do, I'll try Boost.DateTime, and see if it comes out cleaner.